docs(apisix): wrote Metrics with Prometheus (#60)
api7/docs: apisix_versioned_docs/version-3.1.0/how-to-guide/observability/metrics-with-prometheus.md

---
title: Metrics with Prometheus

> slug: /how-to-guide/observabil/metrics-with-prometheus

Reviewer: ```suggestion
slug: /how-to-guide/observability/metrics-with-prometheus
```

apisix_etcd_modify_indexes{key="global_rules"} 0
apisix_etcd_modify_indexes{key="max_modify_index"} 16
apisix_etcd_modify_indexes{key="prev_index"} 15
apisix_etcd_modify_indexes{key="protos"} 0
apisix_etcd_modify_indexes{key="routes"} 16
apisix_etcd_modify_indexes{key="services"} 0
apisix_etcd_modify_indexes{key="ssls"} 0
apisix_etcd_modify_indexes{key="stream_routes"} 0
apisix_etcd_modify_indexes{key="upstreams"} 0
apisix_etcd_modify_indexes{key="x_etcd_index"} 16
# HELP apisix_etcd_reachable Config server etcd reachable from APISIX, 0 is unreachable
# TYPE apisix_etcd_reachable gauge
apisix_etcd_reachable 1
# HELP apisix_http_requests_total The total number of client requests since APISIX started
# TYPE apisix_http_requests_total gauge
apisix_http_requests_total 21
...
```

## Configure Prometheus

Prometheus sees monitored objects as targets. You can configure the APISIX metric endpoint as a target in the Prometheus configuration file `prometheus.yml`.

```shell
echo 'scrape_configs:
  - job_name: "apisix"
    scrape_interval: 15s
    metrics_path: "/apisix/prometheus/metrics"
    static_configs:
      - targets: ["apisix-quickstart:9091"]
' > prometheus.yml
```

> Let's start a Prometheus instance in Docker. The port is mapped to `9092` because `9090` is a reserved port of APISIX. The custom file `prometheus.yml` is mapped to the instance by mounting volume.

Reviewer 2: ```suggestion
Let's start a Prometheus instance in Docker. The exposed port is mapped to `9092` on the host because `9090` is reserved for APISIX. The local configuration file `prometheus.yml` is mounted to the Prometheus container. 
```

```shell
docker run -d --name=apisix-quickstart-grafana \
  -p 3000:3000 \
  --network=apisix-quickstart-net \
  grafana/grafana-oss
```

Add the Prometheus instance created above to Grafana as a data source:

![Grafana Data Source](https://static.apiseven.com/uploads/2023/03/02/E9PNMkdv_grafana-data-source.png)

The official APISIX metric dashboard is published to [Grafana dashboards](https://grafana.com/grafana/dashboards/) with ID [11719](https://grafana.com/grafana/dashboards/11719-apache-apisix/). You can then import the dashboard into Grafana with the ID.

![Import Dashboard](https://static.apiseven.com/uploads/2023/03/02/21YcUlui_grafana-import-dashboard.png)

If everything is OK, the dashboard will automatically visualize metrics in real time.

![Grafana Dashboard](https://static.apiseven.com/uploads/2023/03/02/8hcTkwWW_grafana-dashboard.png)

## Disable Prometheus Plugin

To disable the `prometheus` Plugin, you can set the `_meta.disable` parameter to `true`.

```shell
curl "http://127.0.0.1:9180/apisix/admin/routes/demo-ip" -X PUT -d '
{
  "plugins": {
    "_meta": {
      "disable": true
    }
  },
}'
```

## Next Steps

> You have now learned how to enable APISIX metrics with Prometheus and visualize them in Grafana. Next, you can further explore other resources about observability in this category.

Reviewer 2: ```suggestion
You have now learned how to enable APISIX metrics with Prometheus and visualize them in Grafana. Explore other resources in How-To Guide to customize your APISIX experience. 
```
